diff --git a/hypapps/trustvisor/src/include/tv_log.h b/hypapps/trustvisor/src/include/tv_log.h
index 95011e971..9d2f757f6 100644
--- a/hypapps/trustvisor/src/include/tv_log.h
+++ b/hypapps/trustvisor/src/include/tv_log.h
@@ -48,7 +48,7 @@
 #define TV_LOG_H
 
 #ifndef EU_LOG_LVL
-#define EU_LOG_LVL EU_WARN
+#define EU_LOG_LVL EU_TRACE
 #endif
 
 #ifndef EU_LOG_PREFIX
diff --git a/xmhf/src/libbaremetal/libxmhfutil/include/eulog.h b/xmhf/src/libbaremetal/libxmhfutil/include/eulog.h
index 8416979b4..cdff631e3 100644
--- a/xmhf/src/libbaremetal/libxmhfutil/include/eulog.h
+++ b/xmhf/src/libbaremetal/libxmhfutil/include/eulog.h
@@ -57,7 +57,7 @@
 #endif
 
 #ifndef EU_LOG_LVL
-#define EU_LOG_LVL EU_WARN
+#define EU_LOG_LVL EU_TRACE
 #endif
 
 #ifndef EU_LOG_PRINTLN
